enterprise/tailnet/coordinator.go

```diff
@@ -0,0 +1,507 @@
+package tailnet
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/coderd/database"
+	agpl "github.com/coder/coder/tailnet"
+)
+
+// NewCoordinator creates a new high availability coordinator
+// that uses PostgreSQL pubsub to exchange handshakes.
+func NewCoordinator(logger slog.Logger, pubsub database.Pubsub) (agpl.Coordinator, error) {
+	coord := &haCoordinator{
+		id:                       uuid.New(),
+		log:                      logger,
+		pubsub:                   pubsub,
+		close:                    make(chan struct{}),
+		nodes:                    map[uuid.UUID]*agpl.Node{},
+		agentSockets:             map[uuid.UUID]net.Conn{},
+		agentToConnectionSockets: map[uuid.UUID]map[uuid.UUID]net.Conn{},
+	}
+
+	if err := coord.runPubsub(); err != nil {
+		return nil, xerrors.Errorf("run coordinator pubsub: %w", err)
+	}
+
+	return coord, nil
+}
+
+type haCoordinator struct {
+	id     uuid.UUID
+	log    slog.Logger
+	mutex  sync.RWMutex
```

Comment:
there's an intense amount of locking around this singular mutex. I'm not well-versed in this code that well but if this is supposed to be servicing many requests then this seems like it could slow to a crawl under even moderate load.